battle_strategies.py

```diff
@@ -174,3 +154,106 @@ def pick_move(self, curr_pokemon: Pokemon, opposing_pokemon: Pokemon) -> Move:
+        next_action = self._choose_move_from_policy(next_state, epsilon=False)
+        q_val = self._get_q_value(state, action)
+        td_error = (
+                reward
+                + (self.gamma * self._get_q_value(next_state, next_action))
+                - q_val
+        )
+        self._q_values[(self._extract_state(state), self._extract_action(action))] = q_val + self.alpha * td_error
+
+
+class ApproxQLearningStrategy(BaseQLearningStrategy):
+    def __init__(self, gamma: float, alpha: float, epsilon: float, softmax: bool = False):
+        super().__init__(gamma, alpha, epsilon, softmax)
+        # Initialized for real during training
+        self.weights = defaultdict(lambda: 0.0)
+
+    def load_weights(self, weights: dict[int, float]):
+        self.weights.update(weights)
+
+    def _get_features(self, state: tuple[Pokemon, Pokemon], move: Move):
+        # TODO: Optimize feature extraction
+        pokemon_a, pokemon_b = state
+        type_mod = Type.dmg_modifier(move.info.type, pokemon_b.species.types[0]) * \
+               (Type.dmg_modifier(move.info.type, pokemon_b.species.types[1]) if len(pokemon_b.species.types) > 1 else 1)
+        type_mod /= 4  # Normalize the value between 0 and 1
+
+        dmg_class_mod_stat_atk, dmg_class_mod_stat_def = (pokemon_a.stats.attack, pokemon_b.stats.defense) \
+            if move.info.damage_class is DamageClass.PHYSICAL else (pokemon_a.stats.special, pokemon_b.stats.special)
+        dmg_class_mod = (dmg_class_mod_stat_atk / (dmg_class_mod_stat_atk + dmg_class_mod_stat_def))
+        return [
+            pokemon_a.hp / pokemon_a.stats.total_hp,
+            pokemon_b.hp / pokemon_b.stats.total_hp,
+            dmg_class_mod,
+            type_mod,
+            (move.info.power / 250),
+            (move.info.accuracy if move.info.accuracy is not None else 1),
+            # move.info.hit_info.min_hits,
+            # move.info.hit_info.max_hits,
+            int(move.info.high_crit_ratio),
+            # move.info.priority,
+            move.info.drain,
+            # move.info.healing,
+            # int(move.pp == 0)
+            # TODO: Add more features
+        ]
+
+    def _get_q_value(self, state: tuple[Pokemon, Pokemon], move: Move) -> float:
+        if state[0].fainted:
+            # Losing terminal state
+            return 0
+        features = self._get_features(state, move)
+        q_val = sum(self.weights[ind] * feature for ind, feature in enumerate(features))
+        return q_val
+
+    def _update(self, reward, action: Move, state: tuple[Pokemon, Pokemon], next_state: tuple[Pokemon, Pokemon]):
+        next_action = self._choose_move_from_policy(next_state, epsilon=False)
+        q_val = self._get_q_value(state, action)
+        td_error = (
+                reward
+                + (self.gamma * self._get_q_value(next_state, next_action))
+                - q_val
+        )
+        features = self._get_features(state, action)
```

Comment:
The features are calculated twice during this function execution. Do you think it would be worth caching them to improve performance a bit?